### PR TITLE
[ONNX] Non-global diagnostic context

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100407
* #100299
* __->__ #100219
* #99944
* #99668

Summary:

* `dynamo_export`, and everything within now access diagnostic context through a maintained local
  variable, instead of global.
* Refactored `diagnose_call` decorator to require local diagnostic context, instead of accessing global.
* Modified `test_fx_to_onnx_*.py` tests to only log '*.sarif' logs when `verbose=True`.
* Temporarily removed diagnostics for `OnnxFunction`, as they don't have access to diagnostic context
  anymore. These diagnostics will be the responsibility of `onnxscript`, and they will return once 
  diagnostics system is integrated there.

### CODE REVIEW

File: torch/onnx/_internal/fx/passes/fx_to_onnxscript.py [hunk 1]
```diff
@@ -7,13 +7,14 @@
...
 import types
 import warnings
-from types import FunctionType
 
 from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
 import onnxscript  # type: ignore[import]
 from onnxscript import evaluator, opset18  # type: ignore[import]
-from onnxscript.function_libs.torch_lib import graph_building  # type: ignore[import]
+from onnxscript.function_libs.torch_lib import (  # type: ignore[import]  # type: ignore[attr-defined]
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:15]:
```suggestion
from onnxscript.function_libs.torch_lib import (  # type: ignore[import,attr-defined]
```
Nit

File: torch/onnx/_internal/fx/passes/fx_to_onnxscript.py [hunk 2]
```diff
@@ -26,50 +27,12 @@
...
-    onnxscript.OnnxFunction.__call__
-)
-
-
 @_beartype.beartype
 def _fx_node_to_onnx_message_formatter(
     fn: Callable, args: Tuple[Any, ...], kwargs: Dict[str, Any]
 ) -> str:
-    assert len(args) > 0
-    node = args[0]
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:72]: Comment on 1 because it seems tricky?

**Review by justinchuby — APPROVED** [review]

File: torch/onnx/_internal/fx/passes/fx_to_onnxscript.py (hunk 2 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:72]: Sure. I was thinking in another PR to change the signature style along with the infra to 
```
def _fx_node_to_onnx_message_formatter(
    fn: Callable, diagnostic_context: DiagnosticContext, node: torch.fx.Node, *args, **kwargs
) -> str:
```
wdyt?

Comment (justinchuby) [reply torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:72]: Looking good!

**Pull request closed without merging**